lib/linklist.h

```diff
@@ -288,6 +293,37 @@ extern void list_delete_node(struct list *list, struct listnode *node);
  */
 extern void list_add_list(struct list *list, struct list *add);
 
+/*
+ * Delete all nodes which satisfy a condition from a list.
+ * Deletes the node if cond function returns true for the node.
+ * If function ptr passed is NULL, it deletes all nodes
+ *
+ * list
+ *    list to operate on
+ * cond
+ *    function pointer which takes node data as input and return TRUE or FALSE
+ */
+
+extern void list_cond_delete_nodes(struct list *list, cond_del_fn_t  cond);
+
+/*
+ * Insert a new element into a list with insertion sort if there is no
+ * duplicate element present in the list
```

Comment:
This comment is wrong. What the function actually does is sorted insert, if there is no duplicate element present up to the location in the list where the sorted element will be inserted. Which is behavior that I think a newcomer to this API would find nebulous at best. For instance if `list` is:

```
A -> D -> B -> A
```
and you do `listnode_add_sort_nodup(list, B)` then you get
```
A -> B -> D -> B -> A
```
So your invariant only holds if the entire list is sorted, which is confusing, and in particular isn't documented anywhere.

Reply:
And to add to this, if you're assuming the list is sorted, then you can use a binary insertion sort, which is `O(n log n)`

Reply:
Comment changed accordingly